tracer/src/Datadog.Trace/AppSec/Security.cs

```diff
@@ -267,7 +271,31 @@ private static void AddAppsecSpecificInstrumentations()
         public void Dispose()
         {
             _waf?.Dispose();
-            _rateLimiter.Dispose();
+            _cancellationTokenSource?.Cancel();
+            _cancellationTokenSource?.Dispose();
+        }
+
+        private async void RunRateLimitLoop(TimeSpan period, CancellationToken cancellationToken)
+        {
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    await Task.Delay(period, cancellationToken).ConfigureAwait(false);
+                    _rateLimiter.Reset();
+                }
+            }
+            catch (TaskCanceledException ex)
+            {
+                Log.Debug(ex, "Expected TaskCanceledException");
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error in RunRateLimitLoop");
+                // we could consider trying to reactive the loop in case on an exception,
+                // however not call reset just means spans will no loner be marked with UserKeep
```

Comment:
```suggestion
                // we could consider trying to reactivate the loop in case on an exception,
                // however not call reset just means spans will no longer be marked with UserKeep
```